### PR TITLE
Respect pivot strategy (if given) in `cholesky`

### PR DESCRIPTION
**Pull request by dkarrasch**
When doing `cholesky(::[Wrapped]SparseMatrixCSC, NoPivot())` the pivoting strategy (the default) was not forwarded to the next call (which used to be `cholesky!`, but now on master has another layer to allow SparseArrays.jl to redirect to SuiteSparse's `cholesky`). Even without the wrapper on `A`, the generic method would take over and direct to the generic, highly unperformant in this case, implementation. With the new redirection layer, this does direct to SparseArrays/SuiteSparse, but because the pivot argument was swallowed in the call, the pivoting strategy is no longer respected: SuiteSparse only knows a fill reducing permutation/pivoting strategy.

Long story short: for generic code that might handle (wrapped) sparse matrices, we do not want to allow calls like `cholesky(A, NoPivot())`, but only `cholesky(A)`.

### CODE REVIEW

**Comment by dkarrasch** [comment]
@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/51245#issuecomment-1711751459) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/878e487_vs_5e85876/report.html) is available.

**Comment by dkarrasch** [comment]
This appears to create an issue with Zygote (with the same error type as in , which seems to be very hard to solve?). I have absolutely no idea what needs to be done here or there to fix this. Reaching out to @willtebbutt @oxinabox for some advice.

@nanosoldier `runtests(["PlutoStaticHTML", "EarlyStopping", "Stheno", "ConjugateComputationVI", "MbedTLS", "BayesianLinearRegressors", "GraphQLClient", "OptimKit", "Intervals", "RAPIDS", "LinearMixingModels", "BM3DDenoise", "VoronoiGraph", "DiffEqDevTools", "PerformanceProfilingHttpEndpoints", "Anatta", "Distances", "Polymake"])`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/51245#issuecomment-1712555215) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/878e487_vs_5e85876/report.html) is available.

**Comment by dkarrasch** [comment]
@devmotion Do you have an idea how to solve the AD issues this causes? Or who to ask?

**Comment by devmotion** [comment]
ChainRules-based AD of `cholesky` relies on (and only works for) calls with signature `cholesky(::AbstractMatrix, ::NoPivot)` (and `cholesky(::AbstractMatrix, ::Val{false})` on older Julia versions): https://github.com/JuliaDiff/ChainRules.jl/blob/ba52ec89ddd97a07e79cc35a9fa39019915d203b/src/rulesets/LinearAlgebra/factorization.jl#L467 https://github.com/JuliaDiff/ChainRules.jl/blob/ba52ec89ddd97a07e79cc35a9fa39019915d203b/src/rulesets/LinearAlgebra/factorization.jl#L476-L481 https://github.com/JuliaDiff/ChainRules.jl/blob/ba52ec89ddd97a07e79cc35a9fa39019915d203b/src/rulesets/LinearAlgebra/factorization.jl#L491-L496

If I'm not mistaken with this PR, however, it would not be sufficient anymore to catch the two-argument version but one would have to handle both the two- and one-argument version since `cholesky(A)` is directly forwarded to `_cholesky(cholcopy(A))`. This would be fixable by adding `rrule`s for the one-argument version as well (forwarding to the two-argument version) but it seems a bit unfortunate, I have the feeling that the dispatch structure was a bit clearer before when everything went through the two-argument version.

**Comment by dkarrasch** [comment]
So, for sparse arrays AD of `cholesky` is already giving wrong results? I mean, sparse cholesky does apply a pivoting strategy, which is not `NoPivot`, and it already has a one-arg method, which is not forwarded to some two-arg method.

**Comment by devmotion** [comment]
Sparse arrays are generally only "sparsely" supported in AD, there are many more things to add than just `cholesky` (eg a quick search reveals https://github.com/JuliaDiff/ChainRules.jl/pull/740, https://github.com/JuliaDiff/ChainRules.jl/issues/731, https://github.com/JuliaDiff/ChainRules.jl/issues/468, https://github.com/JuliaDiff/ChainRules.jl/issues/382).

**Comment by DilumAluthge** [comment]
We have moved the LinearAlgebra stdlib to an external repo: https://github.com/JuliaLang/LinearAlgebra.jl

@dkarrasch If you think that this PR is still relevant, please open a new PR on the [LinearAlgebra.jl](https://github.com/JuliaLang/LinearAlgebra.jl) repo.
